File: client/src/lib/messages.ts
export type WaterfallMessage = {
  type: 'waterfall'
  timestamp: number
  freq_min: number
  freq_max: number
  data: string // base64
}

export type DecodedEntry = {
  snr: number
  dt: number
  freq: number
  message: string
}

export type DecodeMessage = {
  type: 'decode'
  period: number
  messages: DecodedEntry[]
}

export type RadioStatusMessage = {
  type: 'radio_status'
  connected: boolean
  freq: number
  mode: string
  ptt: boolean
}

export type QsoStateValue =
  | { state: 'idle' }
  | { state: 'calling_cq'; my_call: string; my_grid: string; tx_freq: number }
  | {
      state: 'in_qso'
      their_call: string
      their_grid: string | null
      their_report: number | null
      my_report: number | null
      step: 'sent_grid' | 'sent_report' | 'sent_roger_report' | 'sent_rr73' | 'sent_73'
      tx_freq: number
    }
  | { state: 'complete'; their_call: string; their_report: number | null; my_report: number | null }

export type QsoUpdateMessage = {
  type: 'qso_update'
  state: QsoStateValue
  next_tx: string | null
  tx_enabled: boolean
  tx_queued: boolean
}

export type HelloMessage = {
  type: 'hello'
  needs_viewer_auth: boolean
  callsign: string
  grid: string
  log_file: string
  rig_host: string
  rig_port: number
  needs_setup: boolean
  os_type: string
  hamlib_available: boolean
}

export type AuthResultMessage = {
  type: 'auth_result'
  success: boolean
}

export type OperatorStatusMessage = {
  type: 'operator_status'
  operator_client_id: string | null
  you_are_operator: boolean
  client_count: number
}

export type LogEntryMessage = {
  type: 'log_entry'
  their_call: string
  their_grid: string | null
  rst_sent: string
  rst_rcvd: string
  freq_hz: number
  band: string
  date: string
  time_on: string
}

export type DeviceListMessage = {
  type: 'device_list'
  inputs: string[]
  outputs: string[]
}

export type ConfigUpdateResultMessage = {
  type: 'config_update_result'
  success: boolean
  message: string | null
  requires_restart: boolean
}

export type RigctldTestResultMessage = {
  type: 'rigctld_test_result'
  success: boolean
  message: string
}

export type SerialPortListMessage = {
  type: 'serial_port_list'
  ports: string[]
}

export type ServerMessage =
  | { type: 'echo'; payload: unknown }
  | { type: 'error'; message: string }
  | HelloMessage
  | AuthResultMessage
  | OperatorStatusMessage
  | WaterfallMessage
  | DecodeMessage
  | RadioStatusMessage
  | QsoUpdateMessage
  | LogEntryMessage
  | DeviceListMessage
  | ConfigUpdateResultMessage
  | RigctldTestResultMessage
  | SerialPortListMessage

export type CompleteSetupPayload = {
  callsign: string
  grid: string
  operator_password: string
  input_device: string | null
  output_device: string | null
  radio_backend: string
  rigctld_host: string
  rigctld_port: number
  rig_model: number | null
  serial_port: string | null
  baud_rate: number | null
}

export type ClientMessage =
  | { type: 'ping' }
  | { type: 'auth'; password: string }
  | { type: 'claim_operator'; password: string }
  | { type: 'release_operator' }
  | { type: 'set_frequency'; freq: number }
  | { type: 'set_mode'; mode: string; passband: number }
  | { type: 'call_cq'; freq: number }
  | { type: 'respond_to'; their_call: string; their_freq: number; tx_freq: number }
  | { type: 'queue_tx'; message: string; freq: number }
  | { type: 'halt_tx' }
  | { type: 'enable_tx'; enabled: boolean }
  | { type: 'set_tx_parity'; parity: number }
  | { type: 'reset_qso' }
  | { type: 'config_update'; section: string; values: Record<string, unknown> }
  | { type: 'test_rigctld' }
  | { type: 'get_serial_ports' }
  | ({ type: 'complete_setup' } & CompleteSetupPayload)

File: server/src/web/messages.rs
use serde::{Deserialize, Serialize};

/// Messages sent from Server → Client
#[derive(Serialize, Clone)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ServerMessage {
    Echo { payload: serde_json::Value },
    Error { message: String },
    /// First message sent to every client on connect.
    Hello {
        needs_viewer_auth: bool,
        callsign:          String,
        grid:              String,
        log_file:          String,
        rig_host:          String,
        rig_port:          u16,
        needs_setup:       bool,
        os_type:           String,
        hamlib_available:  bool,
    },
    /// Response to `Auth` (viewer) or indicates auth state.
    AuthResult { success: bool },
    /// Broadcast whenever operator lock changes; personalised per-client.
    OperatorStatus {
        operator_client_id: Option<String>,
        you_are_operator: bool,
        client_count: usize,
    },
    Waterfall {
        timestamp: f64,
        freq_min:  u32,
        freq_max:  u32,
        data:      String, // base64
    },
    Decode {
        period:   u64,
        messages: Vec<DecodedMessageJson>,
    },
    RadioStatus {
        connected: bool,
        freq:      u64,
        mode:      String,
        ptt:       bool,
    },
    /// Full QSO state update (sent after every decode cycle and on TX commands).
    QsoUpdate {
        /// Serialised `QsoState` — tagged by `"state"` field.
        state:      serde_json::Value,
        /// The message text currently queued for the next TX period.
        next_tx:    Option<String>,
        tx_enabled: bool,
        tx_queued:  bool,
    },
    /// Broadcast when a QSO is successfully logged to ADIF.
    LogEntry {
        their_call: String,
        their_grid: Option<String>,
        rst_sent:   String,
        rst_rcvd:   String,
        freq_hz:    u64,
        band:       String,
        date:       String,
        time_on:    String,
    },
    /// Sent on connect and on demand: available audio devices.
    DeviceList {
        inputs:  Vec<String>,
        outputs: Vec<String>,
    },
    /// Response to a `ConfigUpdate` client message.
    ConfigUpdateResult {
        success:          bool,
        message:          Option<String>,
        requires_restart: bool,
    },
    /// Response to a `TestRigctld` client message.
    RigctldTestResult {
        success: bool,
        message: String,
    },
    /// Response to a `GetSerialPorts` client message.
    SerialPortList { ports: Vec<String> },
}

#[derive(Serialize, Clone)]
pub struct DecodedMessageJson {
    pub snr:     i32,
    pub dt:      f32,
    pub freq:    f32,
    pub message: String,
}

/// Messages sent from Client → Server
#[derive(Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ClientMessage {
    Ping {},
    /// Viewer password authentication (only required if `needs_viewer_auth` was true).
    Auth { password: String },
    /// Claim the operator lock.
    ClaimOperator { password: String },
    /// Release the operator lock.
    ReleaseOperator {},

    SetFrequency { freq: u64 },
    SetMode { mode: String, passband: i32 },

    /// Begin calling CQ on `freq` Hz (audio frequency within passband).
    CallCq { freq: f32 },
    /// Respond to a decoded CQ from `their_call` at `their_freq`.
    RespondTo { their_call: String, their_freq: f32, tx_freq: f32 },
    /// Manually queue a custom FT8 message.
    QueueTx { message: String, freq: f32 },
    /// Emergency stop: drop TX queue and deassert PTT.
    HaltTx {},
    /// Enable or disable automatic TX.
    EnableTx { enabled: bool },
    /// Select TX period parity: 0 = even, 1 = odd.
    SetTxParity { parity: u8 },
    /// Reset QSO state to Idle.
    ResetQso {},

    /// Update a config section. Operator-only.
    /// `section`: "station" | "audio" | "radio" | "network"
    /// `values`: map of field name → new value (as JSON strings/numbers)
    ConfigUpdate { section: String, values: serde_json::Value },

    /// Test the configured rigctld connection. Operator-only.
    TestRigctld {},
    /// Request the list of available serial ports (for Hamlib direct setup).
    GetSerialPorts {},
    /// Complete the setup wizard: saves all config sections at once.
    /// Does not require operator lock; works on first run.
    CompleteSetup {
        callsign:          String,
        grid:              String,
        operator_password: String,
        input_device:      Option<String>,
        output_device:     Option<String>,
        radio_backend:     String,
        rigctld_host:      String,
        rigctld_port:      u16,
        rig_model:         Option<i32>,
        serial_port:       Option<String>,
        baud_rate:         Option<u32>,
    },
}
